libbeat/outputs/tidb_slow_query/schema.go

```diff
@@ -0,0 +1,191 @@
+		"Plan_from_cache":               "tinyint(1)",
+		"Plan_from_binding":             "tinyint(1)",
+		"Prepared":                      "tinyint(1)",
+		"KV_total":                      "double",
+		"PD_total":                      "double",
+		"Backoff_total":                 "double",
+		"Write_sql_response_total":      "double",
+		"Plan":                          "longtext",
+		"Plan_digest":                   "varchar(128)",
+		"Backoff_Detail":                "varchar(4096)",
+		"Rocksdb_delete_skipped_count":  "double",
+		"Rocksdb_key_skipped_count":     "double",
+		"Rocksdb_block_cache_hit_count": "double",
+		"Rocksdb_block_read_count":      "double",
+		"Rocksdb_block_read_byte":       "double",
+	}
+	orderedColumn = make([]string, 0, len(slowQuerySQLType))
+	zone, _       = time.LoadLocation("Asia/Shanghai")
+)
+
+func init() {
+	for k := range slowQuerySQLType {
+		orderedColumn = append(orderedColumn, k)
+	}
+}
+
+const dateFormat = "2006-01-02"
+
+func insertStmt(schema, table string) string {
+	cols := make([]string, 0, len(orderedColumn))
+	args := make([]string, 0, len(orderedColumn))
+	for _, c := range orderedColumn {
+		cols = append(cols, quote(c))
+		args = append(args, "?")
+	}
+	buf := new(bytes.Buffer)
+	buf.WriteString(fmt.Sprintf("INSERT INTO %s.%s (", quote(schema), quote(table)))
+	buf.WriteString(strings.Join(cols, ","))
+	buf.WriteString(") ")
+	buf.WriteString("VALUES (")
+	buf.WriteString(strings.Join(args, ","))
+	buf.WriteString(") ")
+	buf.WriteString(";")
+	return buf.String()
+}
+
+func createTableStmt(schema, table string, lessThanPartitions []time.Time) string {
+	buf := new(bytes.Buffer)
+	buf.WriteString(fmt.Sprintf("CREATE TABLE IF NOT EXISTS %s.%s (", quote(schema), quote(table)))
+	buf.WriteString("`id` bigint(20) unsigned not null AUTO_INCREMENT,")
+	for k, v := range slowQuerySQLType {
+		buf.WriteString(fmt.Sprintf("%s %s,", quote(k), v))
+	}
+	buf.WriteString("PRIMARY KEY (`id`,`Time`)")
+	buf.WriteString(") ")
+	buf.WriteString("PARTITION BY RANGE (FLOOR(UNIX_TIMESTAMP(`Time`))) (")
+	for _, p := range lessThanPartitions {
+		unix := p.Unix()
+		buf.WriteString(fmt.Sprintf("PARTITION %s VALUES LESS THAN (%d),", partName(p), unix))
+	}
+	// delete the last ,
+	if len(lessThanPartitions) > 0 {
+		buf.Truncate(buf.Len() - 1)
+	}
+	buf.WriteString(");")
+	return buf.String()
+}
+
+func calculateLessThanPartitionBoundary(t time.Time, step int) []time.Time {
+	parts := make([]time.Time, 0, step)
+	// ensure start of the day at +8:00 timezone
+	y, m, d := t.In(zone).Date()
+	round := time.Date(y, m, d, 0, 0, 0, 0, zone)
+	for i := 1; i <= step; i++ {
+		parts = append(parts, round.Add(time.Duration(i)*24*time.Hour))
+	}
+	return parts
+}
+
+func getPartitionStmt(schema, table string) string {
+	return fmt.Sprintf("SELECT `partition_name` FROM `information_schema`.`partitions` "+
+		"WHERE table_schema='%s' AND table_name='%s' AND `partition_name` IS NOT NULL order by `partition_name` asc",
```

Comment:
This might lead to SQL injections.

Reply:
It dose. However, since table name, which is cluster id, and from k8s namesapce, is controlled by dbaas code, IMHO the risk is relatively low.